style.md

```diff
@@ -887,112 +893,81 @@ if err := foo.Open(); err != nil {
 </td></tr>
 </tbody></table>
 
-If you have an error that clients may need to detect, and you would like to add
-more information to it (e.g., it is not a static string), then you should use a
-custom type.
+For an error with a dynamic string,
+use [`fmt.Errorf`] if the caller does not need to match it,
+and a custom `error` if the caller does need to match it.
 
 <table>
-<thead><tr><th>Bad</th><th>Good</th></tr></thead>
+<thead><tr><th>No error matching</th><th>Error matching</th></tr></thead>
 <tbody>
 <tr><td>
 
 ```go
-func open(file string) error {
-  return fmt.Errorf("file %q not found", file)
-}
-
-func use() {
-  if err := open("testfile.txt"); err != nil {
-    if strings.Contains(err.Error(), "not found") {
-      // handle
-    } else {
-      panic("unknown error")
-    }
-  }
-}
-```
-
-</td><td>
-
-```go
-type errNotFound struct {
-  file string
-}
+// package foo
 
-func (e errNotFound) Error() string {
-  return fmt.Sprintf("file %q not found", e.file)
+func Open(file string) error {
+  return fmt.Errorf("file %q not found", file)
 }
 
-func open(file string) error {
-  return errNotFound{file: file}
-}
+// package bar
 
-func use() {
-  if err := open("testfile.txt"); err != nil {
-    if _, ok := err.(errNotFound); ok {
-      // handle
-    } else {
-      panic("unknown error")
-    }
-  }
+if err := foo.Open("testfile.txt"); err != nil {
+  // Can't handle the error.
+  panic("unknown error")
 }
 ```
 
-</td></tr>
-</tbody></table>
-
-Be careful with exporting custom error types directly since they become part of
-the public API of the package. It is preferable to expose matcher functions to
-check the error instead.
+</td><td>
 
 ```go
 // package foo
 
-type errNotFound struct {
-  file string
-}
-
-func (e errNotFound) Error() string {
-  return fmt.Sprintf("file %q not found", e.file)
+type NotFoundError struct {
+  File string
 }
 
-func IsNotFoundError(err error) bool {
-  _, ok := err.(errNotFound)
-  return ok
+func (e NotFoundError) Error() string {
```

Comment:
Methods to satisfy the error interface should always be [defined on the pointer receiver](https://stackoverflow.com/questions/50333428/custom-errors-in-golang-and-pointer-receivers).

Reply:
I disagree on that as an "always" rule. It's up to you, based on what semantics you want.
That said, I agree that "if unsure, use pointer receiver" is a good rule of thumb.
Since it's difficult to communicate that nuance in the style guide,
we can change the example code to use a pointer receiver without making an explicit statement about it.

Reply:
It's unfortunately more subtle than that -- there are specific technical reasons why error methods defined on value receivers introduce ambiguities that break callers, though I'm having trouble pulling up a good reference...